### ISSUE TITLE
False-positive E0602 with nonlocal

### ISSUE THREAD

**Issue by bersbersbers** [PyCQA/pylint#2049]
### Steps to reproduce
```
import matplotlib.pyplot as plt

def Main():
    def OnClick(event):
        if True:
            nonlocal i
            i += 1
            print(i)

    i = 0
    fig = plt.figure()
    fig.canvas.mpl_connect('button_press_event', OnClick)
    plt.show(block=True)

Main()
```
### Current behavior
E0602:Undefined variable 'i'

### Expected behavior
No error. Code runs fine, each click in the figure prints an increased counter.

### pylint --version output

No config file found, using default configuration
pylint 1.8.4,
astroid 1.6.3
Python 3.6.5 (v3.6.5:f59c0932b4, Mar 28 2018, 17:00:18) [MSC v.1900 64 bit (AMD64)]

**Comment by PCManticore** [comment]
Thanks, I can reproduce this bug!

**Comment by bersbersbers** [comment]
This one is similar, but with a different error:

```
"""foo"""
def somefun():
    """bar"""
    bug2 = True

    def on_click(event):
        """on_click"""
        if event:
            nonlocal bug2
            bug2 = not bug2

    on_click(True)

somefun()

```

**Comment by sushobhit27** [comment]
closed via #2391
